feat: first pass

devtooligan opened the pull request:
This is an initial draft of an implementation for proposed [EIP-5089, Principal Token Standard](https://github.com/ethereum/EIPs/pull/5095).

devtooligan committed:
feat: first pass

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.14;
+
+import {ERC20} from "solmate/src/tokens/ERC20.sol";
```
You can inherit from yield-utils-v2

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,140 @@
+        address from,
+        address to
+    ) public virtual afterMaturity returns (uint256 underlyingAmount) {
+        if (msg.sender != from) {
```
If you use yield-utils you can replace this block by a call to `_decreaseAllowance`

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,140 @@
+            if (allowed != type(uint256).max) allowance[from][msg.sender] = allowed - principalAmount;
+        }
+        // Check for rounding error since we round down in previewRedeem.
+        require((underlyingAmount = previewRedeem(principalAmount)) != 0, "ZERO_ASSETS");
```
There should be the usual external/internal split of functions

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,140 @@
+
+        emit Redeem(from, to, principalAmount);
+
+        underlying.safeTransfer(to, underlyingAmount);
```
If this contract is to be inherited from, this transfer should be abstracted into an internal function, because the 5089 might not hold the underlying itself, and instead be permissioned to draw it from somewhere else (like a Join)

alcueca reviewed (changes requested):
Very good first pass. Take into account that the next step would be to code a fyToken that uses this to comply with 5089.

devtooligan commented:
@alcueca Addressed all comments.  

Questions:
1) Do we really need external/internal for EVERY function or is that overkill?
2) Is it correct that redeem (and previewRedeem) revert prior to maturity but all other functions do not?
3) Is it overkill to explicitly add a @dev tag on every function to indicate if it reverts prior to maturity?

devtooligan committed:
fix: code review comments

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,194 @@
+        return _previewWithdraw(underlyingAmount);
+    }
+
+    /// @dev This doesn't revert if before maturity.
```
It should revert before maturity, though

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,194 @@
+        return _maxWithdraw(owner);
+    }
+
+    /// @dev This doesn't revert if before maturity.
```
It shouldn't revert before maturity, it should return zero instead

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    /// @dev This doesn't revert if before maturity.
+    function _maxRedeem(address owner) internal view virtual returns (uint256 maxPrincipalAmount) {
```
It should revert before maturity, same as `_redeem`. Better to revert in the internal function to ensure behavioural integrity.

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,194 @@
+
+    /// @notice Allows user to simulate redemption of a given amount of principal tokens, inclusive of fees and other
+    /// current block conditions.
+    /// @dev Reverts prior to maturity.
```
Doesn't, right now, but yes it should.

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,194 @@
+        return _withdraw(underlyingAmount, from, to);
+    }
+
+    /// @dev This doesn't revert if before maturity.
```
It should revert, for behavioural integrity. Otherwise inheriting contracts could mess it up easily.

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    /// @notice Burns a calculated amount of principal tokens in exchange for an exact amount of underlying.
+    /// @dev This doesn't revert if before maturity.
```
It should revert.

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,194 @@
+        _transferOut(to, underlyingAmount);
+    }
+
+    function _transferOut(address to, uint256 amount) internal virtual {
```
Natspec, please.

alcueca reviewed (changes requested):
All good once we get the contract to revert properly.

`_redeem`, `_withdraw`, `_previewRedeem` and `_previewWithdraw` should revert before maturity.

`_max*` should return zero.

`_convert*` should use a 1:1 conversion rate.

A @dev tag in each function specifying the behaviour before maturity seems appropriate.

alcueca commented:
Having internal and external for every function does add a lot of boilerplate, but as far as I know the optimizer is not smart enough yet to do something like that, and calling a public function from within your contract is a CALL, not a JUMP, costing some extra gas.

We need the external functions to comply with the ERC, and if any of those need to call others, then we need to split.

By that logic, I think we don't need internal `redeem`, `withdraw` or `max*`, only `preview*` and `convert*`.

devtooligan commented:
@alcueca Ok, sounds good.  Implementing above changes.  

So `deposit` or `mint` is not part of this standard, correct?  Can you explain a little about the thinking there?

Can you expand on the thinking there please. I thought I saw some mention somewhere about tying in with Pool contract buy/sell functions but I can't find that comment now.

devtooligan committed:
fix: code review comments

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,167 @@
+     *****************************************************************************************************************/
+
+    /// @notice Calculates the amount of underlying tokens that would be exchanged for a given amount of principal tokens.
+    /// @dev This doesn't revert if before maturity.
```
```suggestion
    /// @dev Before maturity, it converts to underlying as if at maturity.
```

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    /// @notice Converts a given amount of underlying tokens to principal exclusive of fees.
+    /// @dev This doesn't revert if before maturity.
```
```suggestion
    /// @dev Before maturity, it converts to principal as if at maturity.
```

alcueca reviewed (approved):
Made two suggestions, that's all.

devtooligan committed:
Update src/ERC5089.sol

Co-authored-by: Alberto Cuesta Cañada <38806121+alcueca@users.noreply.github.com>

devtooligan committed:
Update src/ERC5089.sol

Co-authored-by: Alberto Cuesta Cañada <38806121+alcueca@users.noreply.github.com>

alcueca commented on src/ERC5089.sol:
```diff
@@ -0,0 +1,167 @@
+        return _previewRedeem(principalAmount);
+    }
+
+    function _previewRedeem(uint256 principalAmount) internal view virtual returns (uint256 underlyingAmount) {
```
Just an idea, but maybe it should be `_previewRedeem` and `_previewWithdraw` that revert after maturity, and therefore all functions that call them. We still keep the dev comment in the external functions explaining that they revert after maturity.

It feels weird to have that kind of modifiers in internal functions, but I have the feeling that it is like the data integrity patterns. Some of the `redeem` logic, which includes behaviour before and after maturity, seems to be encoded in the `_previewRedeem` function, a bit like leaving math in an internal function. It seems to me that putting the maturity checks in there is correct.

alcueca reviewed (approved):
The implementation is correct and can be merged, however please look at the comment about where to check for maturity.

devtooligan committed:
tweak: update name to 5095

devtooligan committed:
Merge branch 'feat/first-pass' of github.com:yieldprotocol/ERC5089 into feat/first-pass
